File: gloSuite9023/gloSuite/gloEMRAdmin/gloEMRAdmin/frmAddGCUser.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Net
Imports System.Xml
Imports System.Security.Cryptography.X509Certificates
Imports Microsoft.Web.Administration

Public Class frmAddGCUser 'Code Start-Added by kanchan on 20120731 for gloCommunity
    Private _encryptionKey As String = "12345678"

    Private Sub btnCreateUser_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnCreateUser.Click
        Me.Cursor = Cursors.WaitCursor
        Dim objUserManagementService As UserManagement.UserManagementService = Nothing
        Dim objclsgloCommunityUsers As clsgloCommunityUsers = Nothing
        Dim oclsEncryption As clsEncryption = Nothing
        Dim dtClinic As DataTable = Nothing
        Dim ObjGCUSer As clsgloCommunityUsers = Nothing
        Try
            '#Region "Validations"
            Me.Cursor = Cursors.WaitCursor

            ''Install certificate as per staging/production environment.
            If InstallCertificate() = False Then
                MessageBox.Show("Error in installing certificate", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtUserNm.Focus()
                Return
            End If
            ''end

            If String.IsNullOrEmpty(txtUserNm.Text.Trim()) Then
                MessageBox.Show("Please enter User Name", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtUserNm.Focus()
                Return
            End If
            If CheckSpecialCharacters(txtUserNm.Text) = False Then
                MessageBox.Show("Please enter Valid User Name", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtUserNm.Focus()
                Return
            End If
            If String.IsNullOrEmpty(txtPassword.Text.Trim()) Then
                MessageBox.Show("Please enter Password", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtPassword.Focus()
                Return
            End If
            If CheckPassword(txtPassword.Text) = False Then
                txtPassword.Focus()
                Return
            End If
            If String.IsNullOrEmpty(txtConfirmPwd.Text.Trim()) Then
                MessageBox.Show("Please enter Confirm Password", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtConfirmPwd.Focus()
                Return
            End If
            If String.IsNullOrEmpty(txtEmail.Text.Trim()) Then
                MessageBox.Show("Please enter E-mail", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtEmail.Focus()
                Return
            End If
            If CheckEmail(txtEmail.Text) = False Then
                MessageBox.Show("Please enter valid E-mail address", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtEmail.Focus()
                Return
            End If
            If String.IsNullOrEmpty(txtQuestion.Text.Trim()) Then
                MessageBox.Show("Please enter Security Question", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtQuestion.Focus()
                Return
            End If
            If String.IsNullOrEmpty(txtAnswer.Text.Trim()) Then
                MessageBox.Show("Please enter Security Answer", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtAnswer.Focus()
                Return
            End If
            If txtPassword.Text.Trim() <> txtConfirmPwd.Text.Trim() Then
                MessageBox.Show("Password & Confirmation Password must match", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtConfirmPwd.Focus()
                Return
            End If
            '#End Region

            'Add User
            objUserManagementService = New UserManagement.UserManagementService()
            objUserManagementService.Url = mdlGeneral.gstrSharepointSrvNm + "/" + mdlGeneral.gstr_Layouts & "/UserManagementService.asmx"

            oclsEncryption = New clsEncryption()
            Dim _strEncryptPWD As String = oclsEncryption.EncryptToBase64String(txtPassword.Text.Trim(), _encryptionKey)

            ObjGCUSer = New clsgloCommunityUsers()
            dtClinic = ObjGCUSer.getClinicData()
            If IsNothing(dtClinic) Then
                MessageBox.Show("Clinic Data is not available to register User.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                Return
            End If

            If dtClinic.Rows.Count > 0 Then
                Dim strStatus As String = objUserManagementService.CreateUser(txtUserNm.Text, txtPassword.Text, txtEmail.Text, txtQuestion.Text, txtAnswer.Text, True, dtClinic.Rows(0)(1), dtClinic.Rows(0)(2))

                Select Case strStatus
                    Case "DuplicateUserName"
                        MessageBox.Show("Username already exists. Please enter a different user name.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "DuplicateEmail"
                        MessageBox.Show("A username for that e-mail address already exists. Please enter a different e-mail address.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "InvalidPassword"
                        MessageBox.Show("The password provided is invalid. Please enter a valid password value.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "InvalidEmail"
                        MessageBox.Show("The E-mail address provided is invalid. Please check the value and try again.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "InvalidAnswer"
                        MessageBox.Show("The password retrieval answer provided is invalid. Please check the value and try again.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "InvalidQuestion"
                        MessageBox.Show("The password retrieval question provided is invalid. Please check the value and try again.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "InvalidUserName"
                        MessageBox.Show("The user name provided is invalid. Please check the value and try again.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "ProviderError"
                        MessageBox.Show("The authentication provider returned an error. Please verify your entry and try again. If the problem persists, please contact your system administrator.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "UserRejected"
                        MessageBox.Show("The user creation request has been canceled. Please verify your entry and try again. If the problem persists, please contact your system administrator.", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Return

                    Case "Success"
                        If True Then
                            objclsgloCommunityUsers = New clsgloCommunityUsers()
                            objclsgloCommunityUsers.InsertGCUSer(mdlGeneral.gnLoginID, txtUserNm.Text, _strEncryptPWD, gIscommunitystaging)
                            mdlGeneral.gstrGCUserName = txtUserNm.Text
                            mdlGeneral.gstrGCPassword = txtPassword.Text

                            If IsSiteExist() = False Then
                                If CreateSite() = True Then
                                    MessageBox.Show("User has been created successfully", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                                Else
                                    MessageBox.Show("User has been created successfully.\n Please Note: Failed to create site please contact to system administrator.", gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                                End If
                            Else
                                MessageBox.Show("User has been created successfully", gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                            End If
                            Me.Close()
                            Exit Select
                        End If
                End Select
            End If
        Catch generatedExceptionName As Exception
        Finally
            Me.Cursor = Cursors.Default
            If objUserManagementService IsNot Nothing Then
                objUserManagementService.Dispose()
                objUserManagementService = Nothing
            End If
            If objclsgloCommunityUsers IsNot Nothing Then
                objclsgloCommunityUsers = Nothing
            End If
            If oclsEncryption IsNot Nothing Then
                oclsEncryption = Nothing
            End If
            If Not IsNothing(ObjGCUSer) Then
                ObjGCUSer = Nothing
            End If
            dtClinic = Nothing
            Me.Cursor = Cursors.[Default]
        End Try

    End Sub

    Private Function CheckEmail(ByVal EmailAddress As String) As Boolean
        Dim strPattern As String = "^([0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*@([0-9a-zA-Z][-\w]*[0-9a-zA-Z]\.)+[a-zA-Z]{2,9})$"

        If System.Text.RegularExpressions.Regex.IsMatch(EmailAddress, strPattern) Then

            Return True
        End If
        Return False
    End Function
    'code start-Added by kanchan on 20120802 for form based authentication
    Private Sub txtConfirmPwd_Enter(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles txtConfirmPwd.Enter
        If String.IsNullOrEmpty(txtPassword.Text.Trim()) Then
            MessageBox.Show("Please enter Password", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
            txtPassword.Focus()
            Return
        End If
    End Sub
    'Code Start-Added by kanchan on 20120803 for CheckSpecialCharacters
    Private Function CheckSpecialCharacters(ByVal _text As String) As Boolean
        Dim strPattern As String = "^[a-zA-Z0-9_.-]+$"
        If System.Text.RegularExpressions.Regex.IsMatch(_text, strPattern) Then
            Return True
        End If
        Return False
    End Function
    'Code Start-Added by kanchan on 20120803 for CheckPassword
    Private Function CheckPassword(ByVal _text As String) As Boolean
        Dim strPattern As String = String.Empty
        Try
            strPattern = "^.*(?=.{7,}).*$"
            If System.Text.RegularExpressions.Regex.IsMatch(_text, strPattern) Then
                strPattern = "^.*(?=.*[!@#$%^&+=*]).*$"
                If System.Text.RegularExpressions.Regex.IsMatch(_text, strPattern) Then
                    Return True
                Else
                    MessageBox.Show("Password should contain at least 1 special character", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                    Return False
                End If
            Else
                MessageBox.Show("Password should contain atleast 7 characters,which include 1 special character", mdlGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information)
                Return False
            End If
            Return False
        Catch ex As Exception
            Return False
        Finally
            strPattern = String.Empty
        End Try
    End Function

#Region "Create site"
    Private Function CreateSite() As Boolean
        Dim _result As Boolean = False
        Dim WMService As MeetingService.Meetings = Nothing
        Dim ObjGCUSer As clsgloCommunityUsers = Nothing
        Try
            WMService = New MeetingService.Meetings()
            WMService.Url = gstrSharepointSrvNm
            'ConfigurationManager.AppSettings["SharePointSiteAddress"];
            WMService.CookieContainer = New CookieContainer()
            ObjGCUSer = New clsgloCommunityUsers()

            If oFormCookie Is Nothing Then

                WMService.CookieContainer.Add(ObjGCUSer.QueryToSharePoint(mdlGeneral.gstrGCUserName, mdlGeneral.gstrGCPassword))
            Else
                WMService.CookieContainer.Add(oFormCookie)
            End If

            Dim tz As New MeetingService.TimeZoneInf()

            If (WMService.Url.EndsWith("/")) Then
                WMService.Url = WMService.Url.Remove(WMService.Url.Length - 1, 1)
            End If

            WMService.Url = WMService.Url + gstrSharepointSiteNm + "/_vti_bin/meetings.asmx"

            Dim CreateWorkspaceResponse As System.Xml.XmlNode = Nothing

            If (Not String.IsNullOrEmpty(WMService.Url)) Then
                Dim temp As XmlNode = WMService.GetMeetingWorkspaces(False)
                ' "STS#0" for Team site & 1033 for language(Eng-Us)
                CreateWorkspaceResponse = WMService.CreateWorkspace(gstrClinicName.Trim(), "STS#0", System.UInt32.Parse("1033"), tz)

                Dim objList As New gloLists.Lists()

                objList.CookieContainer = New CookieContainer()
                If oFormCookie Is Nothing Then
                    objList.CookieContainer.Add(ObjGCUSer.QueryToSharePoint(mdlGeneral.gstrGCUserName, mdlGeneral.gstrGCPassword))
                Else
                    objList.CookieContainer.Add(oFormCookie)
                End If


                objList.Url = gstrSharepointSrvNm + "/" + gstrSharepointSiteNm & "/" & gstrClinicName.Trim() & "/_vti_bin/lists.asmx"

                objList.AddList("DataConnections", Nothing, 130)
                '130 for DataConnections Library
                Dim xml As String = "<Batch OnError='Continue'>"
                xml += "<List EnableModeration='False'/>"
                xml += "</Batch>"
                Dim listNode As XmlNode = objList.GetList("DataConnections")
                Dim version As XmlNode = listNode.Attributes("Version")
                Dim doc As New XmlDocument()
                doc.LoadXml(xml)
                Dim propertiesNode As XmlNode = doc.SelectSingleNode("//List")
                Dim resultNode As XmlNode = objList.UpdateList("DataConnections", propertiesNode, Nothing, Nothing, Nothing, version.Value)

                objList.AddList("Repository", Nothing, 101)
                '101 for Document Library
                _result = True
            End If
        Catch ex As Exception
            Me.Cursor = Cursors.[Default]
        Finally
            If WMService IsNot Nothing Then
                WMService.Dispose()
            End If
            If ObjGCUSer IsNot Nothing Then
                ObjGCUSer = Nothing
            End If
        End Try
        Return _result
    End Function

    Private Function IsSiteExist() As Boolean
        Dim _result As Boolean = True
        Dim ObjGCUSer As clsgloCommunityUsers = Nothing
        Try
            Dim olst As New gloLists.Lists()

            olst.CookieContainer = New CookieContainer()
            If oFormCookie Is Nothing Then
                ObjGCUSer = New clsgloCommunityUsers()
                olst.CookieContainer.Add(ObjGCUSer.QueryToSharePoint(mdlGeneral.gstrGCUserName, mdlGeneral.gstrGCPassword))
            Else
                olst.CookieContainer.Add(oFormCookie)
            End If

            olst.Url = gstrSharepointSrvNm + "/" + gstrSharepointSiteNm & "/" & gstrClinicName.Trim() & "/_vti_bin/lists.asmx"

            Dim node As XmlNode = olst.GetListCollection()
        Catch ex As Exception
            _result = False
        End Try
        Return _result
    End Function

    Private Function InstallCertificate() As Boolean
        Dim _IsCertInstalled As Boolean = False

        Try
            Dim mgr As ServerManager = Nothing
            Dim server As String = Environment.MachineName
            ' or remote machine name
            If [String].IsNullOrEmpty(server) Then
                mgr = New ServerManager()
            Else
                mgr = ServerManager.OpenRemote(server)
            End If

            'For some reason this created a new store called 'Personal' - we'll have to figure that out
            Dim store As New X509Store(StoreName.Root, StoreLocation.LocalMachine)
            store.Open(OpenFlags.OpenExistingOnly Or OpenFlags.ReadWrite)

            'Looks like we can create this from a byte array as well
            Dim _strgloCommunityCert As String = "StagingCert.cer"
            If gIscommunitystaging = False Then
                _strgloCommunityCert = "ProductionCert.cer"
            End If

            Dim certificate As New X509Certificate2(System.Windows.Forms.Application.StartupPath & "\gloCommunityCertificates\" & _strgloCommunityCert)

            store.Add(certificate)
            store.Close()
            _IsCertInstalled = True
        Catch ex As Exception
            Me.Cursor = Cursors.[Default]
        End Try
        Return _IsCertInstalled
    End Function
#End Region

    Private Sub frmAddGCUser_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        mdlGeneral.Fill_Clinic()
    End Sub
End Class

File: gloSuite9023/gloSuite/gloCommon/gloCommunity/Forms/frmAddGCUser.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using gloCommunity.Classes;
using System.Net;
using System.Xml;
using Microsoft.Web.Administration;
using System.Security.Cryptography.X509Certificates;
namespace gloCommunity.Forms
{
    public partial class frmAddGCUser : Form
    {
        string _encryptionKey = "12345678";
    //    string _strEncryptPWD = "";
        public frmAddGCUser()
        {
            InitializeComponent();
        }

        private void btnCreateUser_Click(object sender, EventArgs e)
        {
            this.Cursor = Cursors.WaitCursor;
            UserManagement.UserManagementService objUserManagementService = null;
            clsgloCommunityUsers objclsgloCommunityUsers = null;
            clsEncryption oclsEncryption = null;
            try
            {
                #region Validations

                //Install certificate as per staging/production environment.
                if (InstallCertificate() == false)
                {
                    MessageBox.Show("Error in installing certificate", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtUserNm.Focus();
                    return;
                }
                //end

                if (string.IsNullOrEmpty(clsGeneral.gstrClinicName.Trim()))
                {
                    MessageBox.Show("Please enter Clinic details from admin", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtUserNm.Focus();
                    return;
                }

                if (string.IsNullOrEmpty(txtUserNm.Text.Trim()))
                {
                    MessageBox.Show("Please enter User Name", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtUserNm.Focus();
                    return;
                }
                if (CheckSpecialCharacters(txtUserNm.Text) == false)
                {
                    MessageBox.Show("Please enter Valid User Name", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtUserNm.Focus();
                    return;
                }

                if (string.IsNullOrEmpty(txtPassword.Text.Trim()))
                {
                    MessageBox.Show("Please enter Password", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtPassword.Focus();
                    return;
                }
                if (CheckPassword(txtPassword.Text.Trim()) == false)
                {
                    txtPassword.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(txtConfirmPwd.Text.Trim()))
                {
                    MessageBox.Show("Please enter Confirm Password", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtConfirmPwd.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(txtEmail.Text.Trim()))
                {
                    MessageBox.Show("Please enter E-mail", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtEmail.Focus();
                    return;
                }
                if (CheckEmail(txtEmail.Text) == false)
                {
                    MessageBox.Show("Please enter valid E-mail address", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtEmail.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(txtQuestion.Text.Trim()))
                {
                    MessageBox.Show("Please enter Security Question", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtQuestion.Focus();
                    return;
                }
                if (string.IsNullOrEmpty(txtAnswer.Text.Trim()))
                {
                    MessageBox.Show("Please enter Security Answer", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtAnswer.Focus();
                    return;
                }
                if (txtPassword.Text.Trim() != txtConfirmPwd.Text.Trim())
                {
                    MessageBox.Show("Password & Confirmation Password must match", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    txtConfirmPwd.Focus();
                    return;
                }
                #endregion

                //Add User
                objUserManagementService = new UserManagement.UserManagementService();
                objUserManagementService.Url = clsGeneral.gstrSharepointSrvNm + "/" + clsGeneral.gstr_Layouts + "/UserManagementService.asmx";

                oclsEncryption = new clsEncryption();
                string _strEncryptPWD = oclsEncryption.EncryptToBase64String(txtPassword.Text.Trim(), _encryptionKey);

                objclsgloCommunityUsers = new clsgloCommunityUsers();
                DataTable dtClinic = objclsgloCommunityUsers.getClinicData();
                if (dtClinic != null && dtClinic.Rows.Count > 0)
                {
                    string strStatus = objUserManagementService.CreateUser(txtUserNm.Text, txtPassword.Text, txtEmail.Text, txtQuestion.Text, txtAnswer.Text, true, dtClinic.Rows[0]["sClinicName"].ToString(), dtClinic.Rows[0]["sExternalcode"].ToString());

                    switch (strStatus)
                    {
                        case "DuplicateUserName":
                            MessageBox.Show("Username already exists. Please enter a different user name.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "DuplicateEmail":
                            MessageBox.Show("A username for that e-mail address already exists. Please enter a different e-mail address.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "InvalidPassword":
                            MessageBox.Show("The password provided is invalid. Please enter a valid password value.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "InvalidEmail":
                            MessageBox.Show("The E-mail address provided is invalid. Please check the value and try again.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "InvalidAnswer":
                            MessageBox.Show("The password retrieval answer provided is invalid. Please check the value and try again.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "InvalidQuestion":
                            MessageBox.Show("The password retrieval question provided is invalid. Please check the value and try again.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "InvalidUserName":
                            MessageBox.Show("The user name provided is invalid. Please check the value and try again.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "ProviderError":
                            MessageBox.Show("The authentication provider returned an error. Please verify your entry and try again. If the problem persists, please contact your system administrator.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "UserRejected":
                            MessageBox.Show("The user creation request has been canceled. Please verify your entry and try again. If the problem persists, please contact your system administrator.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                            return;

                        case "Success":
                            {
                                objclsgloCommunityUsers.InsertGCUSer(clsGeneral.gnLoginID, txtUserNm.Text, _strEncryptPWD, clsGeneral.gblnIscommunityStaging);

                                gloCommunity.Classes.clsgloCommunityUsers clsGCUSer = new gloCommunity.Classes.clsgloCommunityUsers();
                                DataTable dtUSer = clsGCUSer.getGCUser(clsGeneral.gnLoginID, clsGeneral.gblnIscommunityStaging);
                                if ((dtUSer != null) & dtUSer.Rows.Count > 0)
                                {
                                    //'get gloCommunity username & password as per login id

                                    _strEncryptPWD = oclsEncryption.DecryptFromBase64String(dtUSer.Rows[0]["gc_sPassword"].ToString(), _encryptionKey);
                                    gloCommunity.Classes.clsGeneral.gstrGCUserName = dtUSer.Rows[0]["gc_sUserName"].ToString();
                                    gloCommunity.Classes.clsGeneral.gstrGCPassword = _strEncryptPWD;
                                }
                                if (IsSiteExist() == false)//if site not exist then create
                                {

                                    if (CreateSite() == true)
                                        MessageBox.Show("User has been created successfully", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                                    else
                                        MessageBox.Show("User has been created successfully.\n Please Note: Failed to create site please contact to system administrator.", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                                }
                                else
                                    MessageBox.Show("User has been created successfully", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                                this.DialogResult = System.Windows.Forms.DialogResult.OK;
                                this.Close();
                                break;
                            }
                    }
                }
                else
                    MessageBox.Show("Please enter Clinic details from admin", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);

            }
            catch (Exception)
            {
            }
            finally
            {
                if (objUserManagementService != null)
                {
                    objUserManagementService.Dispose();
                    objUserManagementService = null;
                }
                if (objclsgloCommunityUsers != null)
                    objclsgloCommunityUsers = null;
                if (oclsEncryption != null)
                    oclsEncryption = null;
                this.Cursor = Cursors.Default;
            }

        }

        private bool CheckEmail(string EmailAddress)
        {
            string strPattern = "^([0-9a-zA-Z]([-.\\w]*[0-9a-zA-Z])*@([0-9a-zA-Z][-\\w]*[0-9a-zA-Z]\\.)+[a-zA-Z]{2,9})$";

            if (System.Text.RegularExpressions.Regex.IsMatch(EmailAddress, strPattern))

            { return true; }
            return false;
        }

        private bool CheckSpecialCharacters(string _text)
        {
            string strPattern = "^[a-zA-Z0-9_.-]+$";
            if (System.Text.RegularExpressions.Regex.IsMatch(_text, strPattern))
            {
                return true;
            }
            return false;
        }

        private bool CheckPassword(string _text)
        {
            string strPattern = string.Empty;
            try
            {
                strPattern = "^.*(?=.{7,}).*$";
                if (System.Text.RegularExpressions.Regex.IsMatch(_text, strPattern))
                {
                    strPattern = "^.*(?=.*[!@#$%^&+=*]).*$";
                    if (System.Text.RegularExpressions.Regex.IsMatch(_text, strPattern))
                    {
                        return true;
                    }
                    else
                    {
                        MessageBox.Show("Password should contain at least 1 special character", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                        return false;
                    }
                }
                else
                {
                    MessageBox.Show("Password should contain atleast 7 characters,which include 1 special character", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                    return false;
                }
               // return false;
            }
            catch //(Exception ex)
            {
                return false;
            }
            finally
            {
                strPattern = string.Empty;
            }
        }

        private void txtConfirmPwd_Enter(object sender, EventArgs e)
        {
            if (string.IsNullOrEmpty(txtPassword.Text.Trim()))
            {
                MessageBox.Show("Please enter Password", clsGeneral.gstrMessageBoxCaption, MessageBoxButtons.OK, MessageBoxIcon.Information);
                txtPassword.Focus();
                return;
            }

        }

        #region Create Site
        private bool CreateSite()
        {
            bool _result = false;
            MeetingService.Meetings WMService = null;
            try
            {
                WMService = new MeetingService.Meetings();
                WMService.Url = clsGeneral.gstrSharepointSrvNm;//ConfigurationManager.AppSettings["SharePointSiteAddress"];
                WMService.CookieContainer = new CookieContainer();

                if (clsGeneral.oFormCookie == null)
                    WMService.CookieContainer.Add(clsGeneral.QueryToSharePoint(clsGeneral.gstrAuthenticationWSAddress, clsGeneral.gstrGCUserName, clsGeneral.gstrGCPassword));
                else
                    WMService.CookieContainer.Add(clsGeneral.oFormCookie);

                MeetingService.TimeZoneInf tz = new MeetingService.TimeZoneInf();

                if ((WMService.Url.EndsWith("/")))
                {
                    WMService.Url = WMService.Url.Remove(WMService.Url.Length - 1, 1);
                }

                WMService.Url = WMService.Url + clsGeneral.gstrSharepointSiteNm + "/_vti_bin/meetings.asmx";

                System.Xml.XmlNode CreateWorkspaceResponse = default(System.Xml.XmlNode);

                if ((!string.IsNullOrEmpty(WMService.Url)))
                {
                    XmlNode temp = WMService.GetMeetingWorkspaces(false);
                    // "STS#0" for Team site & 1033 for language(Eng-Us)
                    CreateWorkspaceResponse = WMService.CreateWorkspace(clsGeneral.gstrClinicName.Trim(), "STS#0", System.UInt32.Parse("1033"), tz);

                    gloLists.Lists objList = new gloLists.Lists();

                    objList.CookieContainer = new CookieContainer();
                    if (clsGeneral.oFormCookie == null)
                        objList.CookieContainer.Add(clsGeneral.QueryToSharePoint(clsGeneral.gstrAuthenticationWSAddress, clsGeneral.gstrGCUserName, clsGeneral.gstrGCPassword));
                    else
                        objList.CookieContainer.Add(clsGeneral.oFormCookie);


                    objList.Url = clsGeneral.gstrSharepointSrvNm + "/" + clsGeneral.gstrSharepointSiteNm + "/" + clsGeneral.gstrClinicName.Trim() + "/_vti_bin/lists.asmx";

                    objList.AddList("DataConnections", null, 130);//130 for DataConnections Library
                    string xml = "<Batch OnError='Continue'>";
                    xml += "<List EnableModeration='False'/>";
                    xml += "</Batch>";
                    XmlNode listNode = objList.GetList("DataConnections");
                    XmlNode version = listNode.Attributes["Version"];
                    XmlDocument doc = new XmlDocument();
                    doc.LoadXml(xml);
                    XmlNode propertiesNode = doc.SelectSingleNode("//List");
                    XmlNode resultNode = objList.UpdateList("DataConnections", propertiesNode, null, null, null, version.Value);

                    objList.AddList("Repository", null, 101);//101 for Document Library
                    _result = true;
                }
            }
            catch //(Exception ex)
            {
                this.Cursor = Cursors.Default;
            }
            finally
            {
                if (WMService != null)
                {
                    WMService.Dispose();
                }
            }
            return _result;
        }

        private bool IsSiteExist()
        {
            bool _result = true;
            try
            {
                gloLists.Lists olst = new gloLists.Lists();

                olst.CookieContainer = new CookieContainer();
                if (clsGeneral.oFormCookie == null)
                    olst.CookieContainer.Add(clsGeneral.QueryToSharePoint(clsGeneral.gstrAuthenticationWSAddress, clsGeneral.gstrGCUserName, clsGeneral.gstrGCPassword));
                else
                    olst.CookieContainer.Add(clsGeneral.oFormCookie);

                olst.Url = clsGeneral.gstrSharepointSrvNm + "/" + clsGeneral.gstrSharepointSiteNm + "/" + clsGeneral.gstrClinicName.Trim() + "/_vti_bin/lists.asmx";
                XmlNode node = olst.GetListCollection();

            }
            catch //(Exception ex)
            {
                _result = false;
            }
            return _result;
        }

        private bool InstallCertificate()
        {
            bool _IsCertInstalled = false;

            try
            {
                ServerManager mgr = null;
                string server = Environment.MachineName; // or remote machine name
                if (String.IsNullOrEmpty(server))
                {
                    mgr = new ServerManager();
                }
                else
                {
                    mgr = ServerManager.OpenRemote(server);
                }

                //For some reason this created a new store called 'Personal' - we'll have to figure that out
                X509Store store = new X509Store(StoreName.Root, StoreLocation.LocalMachine);
                store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadWrite);

                //Looks like we can create this from a byte array as well
                string _strgloCommunityCert = "StagingCert.cer";
                if (clsGeneral.gblnIscommunityStaging == false)
                    _strgloCommunityCert = "ProductionCert.cer";

                X509Certificate2 certificate = new X509Certificate2(System.Windows.Forms.Application.StartupPath + "\\gloCommunityCertificates\\" + _strgloCommunityCert);

                store.Add(certificate);
                //object thumb = "bf 08 fc e1 13 9e 96 0e b0 43 8d 07 92 97 5c 5a c0 59 e6 19";
                //X509Certificate2Collection test = store.Certificates.Find(X509FindType.FindByThumbprint, thumb, false);
                store.Close();
                _IsCertInstalled = true;
            }
            catch //(Exception ex)
            {
                this.Cursor = Cursors.Default;
            }
            return _IsCertInstalled;
        }
        #endregion
    }
}
